daft/internal/gpu.py

```diff
@@ -1,18 +1,44 @@
 from __future__ import annotations
 
-import subprocess
 
-from daft.dependencies import ET
+def _raw_device_count_nvml() -> int:
+    """
+    Return number of devices as reported by NVML or zero if NVML discovery/initialization failed.
 
+    Inspired by PyTorch: https://github.com/pytorch/pytorch/blob/88e54de21976aa504e797e47f06b480b9108ef5c/torch/cuda/__init__.py#L711
+    """
+    from ctypes import CDLL, byref, c_int
 
-def cuda_device_count():
-    """Returns the number of CUDA devices detected by nvidia-smi command"""
     try:
-        nvidia_smi_output = subprocess.check_output(["nvidia-smi", "-x", "-q"])
-    except Exception:
+        nvml_h = CDLL("libnvidia-ml.so.1")
```

Comment:
In my testing, `nvidia-smi` did not seem to respect `CUDA_VISIBLE_DEVICES` but that might be something I'm doing wrong since most people online report that it does.

The reason I have this is actually that the GPU numbers outputted by `nvidia-smi` may not actually map to the correct device numbers: https://forums.developer.nvidia.com/t/cuda-visible-devices-being-ignored/41808/5

As for the xml library, I believe it's actually part of the python std lib.